[IR] Added different suffix to fresh names between MacroAST and JavaAST.

ggevay opened the pull request:
Also, removed etaCompact, etaExpand, TermName.Eta, TermName.While, TermName.DoWhile.

Fixes #265.

ggevay commented:
> We should avoid ending up with names such as foo$m01$m02$m03

But the suffixes getting concatenated might be useful for tracking where did the name come from. E.g. if I'm printing all the trees after every transformation step, and I see `x$42$3` in a large tree, then I can look for `x$42` in the preceding tree.

joroKr21 commented on emma-language/src/main/scala/org/emmalanguage/ast/AST.scala:
```diff
@@ -183,4 +183,6 @@ trait AST extends CommonAST
     dfn <- defs.tail
   } yield dfn
+
+  private[ast] def freshNameSuffix: String
```
:+1: for returning a `Char`. If a more complex suffix is needed later, changing the type won't affect use sites.

aalexandrov commented:
> But the suffixes getting concatenated might be useful for tracking where did the name come from. E.g. if I'm printing all the trees after every transformation step, and I see `x$42$3` in a large tree, then I can look for `x$42` in the preceding tree.

I agreed, but reading pretty printed code which has 2-3 suffixes for a variable named `x` can make your head hurt very fast. Maybe there should be a field which should serve as a Boolean option indicating whether the `fresh`  should keep input suffix (per default should be set to `false`).

ggevay commented:
I've addressed all the comments.

About avoiding multiple suffixes, we discussed it offline with @aalexandrov, and concluded that it's not that important.

aalexandrov commented:
D'oh, clicking "update branch" after merging another PR creates a merge commit.

aalexandrov commented:
@joroKr21 I think that once you approve you can also merge this.
